Report faulty profiles instead of crashing

james-stocks opened the pull request:
Fixes #5107 

When InSpec cannot load a profile, it should log the failed profile instead of terminating InSpec entirely.

Aha! Link: https://chef.aha.io/features/SH-2553

Commit:
Set profiles that fail to load as failed

Signed-off-by: James Stocks <jstocks@chef.io>

Commit:
Change profile skip_message to status_message

Keep a duplicate of the message in skip_message for the JSON reporters, for backwards compaibility.

Signed-off-by: James Stocks <jstocks@chef.io>

james-stocks commented on lib/inspec/runner_rspec.rb:
```diff
@@ -90,9 +90,12 @@ def exit_code
 
       stats = @formatter.results[:statistics][:controls]
+      load_failures = @formatter.results[:profiles]&.select { |p| p[:status] == "failed" }&.any?
       skipped = @formatter.results.dig(:profiles, 0, :status) == "skipped"
```
I noticed this line _might_ be wrong, as it requires the first (`0`) profile to be skipped; when the run might have a passing profile ahead of the skipped profile.

james-stocks commented on lib/inspec/reporters/json.rb:
```diff
@@ -31,22 +31,22 @@ def platform
-          message:      r[:message],
-          exception:    r[:exception],
-          backtrace:    r[:backtrace],
+          status:         r[:status],
```
I need to back this out, this spacing change is a mistake.

james-stocks commented on test/functional/inspec_exec_test.rb:
```diff
@@ -33,11 +33,12 @@ def stderr
 
-    _(stdout).must_be_empty
-    _(stderr).must_include "Unsupported resource/backend combination: file / aws. Exiting."
+    _(stderr).must_be_empty
```
This test is being changed because the previous version would now be counted as a "failed to load" profile. And that's not the point of this test. It now asserts that the expected error message appears on stdout (as it does as a test failure with the CLI reporter).

Commit:
Address linting mistakes.

Signed-off-by: James Stocks <jstocks@chef.io>

Commit:
Add profile load failure message to CLI reporter

Signed-off-by: James Stocks <jstocks@chef.io>

clintoncwolfe commented on lib/inspec/profile.rb:
```diff
@@ -197,7 +202,7 @@ def params
 
     def collect_tests(include_list = @controls)
-      unless @tests_collected
+      unless @tests_collected || @failed
```
Maybe use `failed?` here

james-stocks replied:
Good call, done.

clintoncwolfe commented on lib/inspec/profile.rb:
```diff
@@ -94,6 +94,7 @@ def initialize(source_reader, options = {})
       @input_values = options[:inputs]
       @tests_collected = false
       @libraries_loaded = false
+      @failed = false
```
It appears that status (used for skipped) is a property of metadata, while failed status is being represented as an instance variable directly on Profile. Could failed be a value of the metadata property?

james-stocks replied:
I think you're right that having `failed` as an instance variable might not be the right place.
However, I'm not seeing an existing `status` in the metadata. The `skipped` status is instead inferred by the profile by calling [`supports_platform?`](https://github.com/inspec/inspec/blob/master/lib/inspec/profile.rb#L327) on the metadata instance.

How would you see this new `failed` property neatly fitting into the metadata? It could be a new `status` metadata property, but I'd need to rework how skipped profiles work as well.

clintoncwolfe replied:
This is an awkward location, but there is a status on the dependency structure, which is set in a really weird way. 
https://github.com/inspec/inspec/pull/5128/files#diff-d1eede4085683a605d9b8892a198f642R265

james-stocks replied:
I did see that, but that is metadata for dependency profiles, not metadata for the immediate profile, no? Or can it be used to represent the metadata for the current profile?
At this point I'd rather just have the instance variable than a metadata entry that behaves weirdly.

clintoncwolfe replied:
I agree the existing approach on Metadata's dependency structure is inappropriate.  

What I'm saying is that the responsibility for knowing the status of the profile is currently on the Metadata object. This PR now divides that responsibility between the Profile and Metadata objects, which creates a lot of potential for bugs.  An implementation could be chosen - an instance variable on Metadata, perhaps - that keeps the responsibility on one class. 

You could still have `Profile#failed?` - it can just call into metadata.

james-stocks replied:
I'll have a go at this, but I might need to leave a backlogged issue to fix how skipping works so that it is also a real status and not inferred indirectly.

james-stocks replied:
Looking at this a second time, `Metadata` is intended to represent the static metadata stored in `my_inspec_profile/metadata.rb`

https://github.com/inspec/inspec/blob/master/lib/inspec/metadata.rb#L12

It does not appear to be the right place to represent an instance of the profile loaded by the runner - this file (`lib/inspec/profile.rb`) is the correct place for that.

So I would like to keep the change here. What should change is to have the instance variable be `@state` not `@failed` and have it store whether this instance of the profile is `:loaded` `:failed` or `:skipped` (the latter would mean directly storing here if this instance of the profile is skipped instead of calling into the metadata).

@clintoncwolfe if you don't agree with me, can we at least agree on a refactoring issue to get the responsibility of Metadata fixed up and documented properly outside of this PR?

clintoncwolfe replied:
Your approach sounds reasonable - having an ivar named state on Profile seems fine, and (ideally) moving the skipped status to be indicated using that ivar sounds good. Adding a comment to the effect that Metadata represents metadata and Profile represents state might be helpful, as they have some overlap. Up to you if you want that on a separate PR.

Schwad replied:
Just had a look at this and appreciate the thought gone into this issue on the backscroll. Looking at this from a newer set of eyes I do appreciate the clarifying comment on metadata (on an initial reading of the object class my interpretation would have been that it solely holds responsibility for `metadata.rb`). 

I also found it useful to clarify that `Profile` is the runtime state representation, and the use of `@state` seems fitting to me. (Naming this is hard, no?) Thanks for pulling this together.

clintoncwolfe commented on lib/inspec/run_data.rb:
```diff
@@ -47,7 +47,7 @@ class RunData
     # and new Hash compatibility behavior will be added during the core reporter plugin
     # conversion process.
-    SCHEMA_VERSION = "0.1.0".freeze
+    SCHEMA_VERSION = "0.2.0".freeze
```
The rundata docs need to be updated - https://github.com/inspec/inspec/blob/94c30647ed8a3374f576379553a73ae7e1587a4c/docs/dev/plugins.md#L449

james-stocks replied:
Good call, thanks, I pushed an update to this file.

clintoncwolfe commented on lib/inspec/schema/exec_json.rb:
```diff
@@ -100,10 +100,10 @@ module ExecJson
+          # These are generated at runtime, and all except status_message are guaranteed
           "sha256" => Primitives::STRING,
           "status" => Primitives::STRING,
-          "skip_message" => Primitives::STRING, # If skipped, why
```
Shouldn't this be retained for backwards compatibility? My understanding is that on a skip, both skip_message and status_message will be present.

james-stocks replied:
I'm taking advantage of the API being pre-1.0 to just drop it. `status_message` should be used. I'm adding `skip_message` in the reporters until 5.0 when they can be cut out.

clintoncwolfe replied:
This isn't part of the (internal) pre-1.0 rundata API though; this is part of the published schema for the (external) output of the JSON reporter. This will impact people.

james-stocks replied:
Ah, people running JSON profiles generated by older InSpec versions? I'll put `skip_message` back in then

james-stocks replied:
I put it back but changed the comment.

Schwad replied:
If this was re-added, would we need it to be added in `plugins.md` or left out due to deprecation?

james-stocks replied:
`plugins.md` documents the run data API and this isn't that.

clintoncwolfe commented on lib/inspec/schema.rb:
```diff
@@ -147,7 +147,7 @@ class Schema
         "license" => { "type" => "string", "optional" => true },
         "summary" => { "type" => "string", "optional" => true },
         "status" => { "type" => "string", "optional" => false },
-        "skip_message" => { "type" => "string", "optional" => true },
```
I suspect this should be retained; see other comment.

james-stocks replied:
I trust you're right and I'll put it back, but I'd like to know - what's the use case where this breaks for users?

clintoncwolfe replied:
(This has nothing to do with the internal RunData API)

This is part of the schema components used to construct the schemae for the JSON output of InSpec - both the JSON profile export and the JSON reporter. We don't know how much the schema are used, but since we are retaining the fields in the output in the JSON reporter when the profile is skipped, if we drop the skip_message field from the schema, the schema will definitely be wrong.

People definitely use the JSON profile export and JSON reporter; we don't know if users are validating against the schema but breaking that would be bad.

james-stocks replied:
Thanks very much for explaining - I need to learn more about these bits of InSpec.

I put `skip_message` back in with a comment that `status_message` should now be used for skip reasons.

Schwad replied:
Status_message is a great approach but definitely appreciate Clinton's comments on how this affects downstream users. Would this be a candidate for an InSpec 5 deprecation? (either one to add then or add now for future removal?) 

But before going that far, is there definitely no edge case in which we'd want both a skip_message and a status_message? 🤔  my initial reaction is no; because if something is skipped in theory there ought not to be anymore to say on the subject.

james-stocks replied:
I don't think there's an opportunity here to add a deprecation warning since this is just a schema. We can just add it to our list of things to take care of when cutting InSpec 5.0.

I don't foresee a case where 2 statuses are needed. At some point we potentially might need to add code to decide which of 2 statuses trumps the other. But currently once it reaches `failed` or `skipped` state, that should be the profile's final outcome.

clintoncwolfe reviewed (commented):
This looks great overall, I especially like the testing. I'm not sure the failed status is being stored in the right place, and there is a missing doc entry, but I think this is pretty close.

Schwad commented on lib/inspec/profile.rb:
```diff
@@ -206,7 +211,12 @@ def collect_tests(include_list = @controls)
           abs_path = source_reader.target.abs_path(path)
-          @runner_context.load_control_file(content, abs_path, nil)
+          begin
+            @runner_context.load_control_file(content, abs_path, nil)
```
Thanks for doing this! So I'm guessing the flow here is during the `collect_tests` phase for the profile object, this is where you check and kill things if `load_control_file` fails, to isolate and document that failure against the Profile

Schwad commented on lib/inspec/runner.rb:
```diff
@@ -115,8 +115,14 @@ def load
+          tests = profile.collect_tests
+          all_controls += tests unless tests.nil?
+        rescue Inspec::Exceptions::ProfileLoadFailed => e
+          Inspec::Log.error "Failed to load profile #{profile.name}: #{e}"
```
....and here's where it seems we catch the `raise` referenced earlier and set our message. Nice! 👍

Schwad commented on test/fixtures/profiles/raise_outside_control/controls/raises.rb:
```diff
@@ -0,0 +1,11 @@
+# While this is a user mistake, users may not know about it if we don't report it...
+
+raise StandardError, 'Something unforeseen...'
+
```
Would this feature also work for a `raise` inside of a control block but outside of a `describe` block? I feel like that code also gets evaluated but it also might be a case you're not looking to solve with this issue.

james-stocks replied:
See the issue: https://github.com/inspec/inspec/issues/5107#issuecomment-648726990

There's at least one other case that needs a different fix, it will be filed as a separate issue.

Schwad reviewed (commented):
Apologies, my two earlier comments I mashed the "comment" button instead of bundling them into a review so now you've probably got extra notifications. Those two comments were just me making sure I understood the flow for solving this correctly. 

Great stuff and looks good to me, thanks for pulling this together! 👍  I also like how you created a home for `status_message` which can be used in new ways.

Commit:
Improve profile status rendering in html2 reporter

Signed-off-by: James Stocks <jstocks@chef.io>

Commit:
Make profile.rb own profile state

Add a @state ivar to the Profile class to represent state.
Add additional comments to the Metadata class to clarify that it represents metadata but not runtime state.

Signed-off-by: James Stocks <jstocks@chef.io>

Schwad reviewed (approved):
Thank you for doing this! 🚀  LGTM 👍

clintoncwolfe reviewed (approved):
Thanks!
